{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preparation of ChAR-seq files for analysis in python with `chartools`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we generate the necessary files to ingest the ChAR-seq data into . The starting point are dna.bed.gz and rna.bed.gz files generated by the snakemake preprocessing pipeline. \n",
    "\n",
    "The path of all the snakemake pipeline output files and of the files generated by this script is specific to the Sherlock computer cluster where the analysis was run. Users will need to adjust these paths according to their system. \n",
    "\n",
    "Required input files are .rna.bed.gz and .dna.bed.gz for all samples. \n",
    "The samples processed here are named as follow:\n",
    "- ES rep 1: TFONM2_ES\n",
    "- ES rep 2: TMF2_Rep2B_ES\n",
    "- DE rep 1: TGNQ5_DE\n",
    "- DE rep 2: TMF3_Rep2A_DE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleanup multimappers and ambiguous reads\n",
    "\n",
    "\n",
    "We first filter the rna.bed.gz and dna.bed.gz to remove reads that map to more that one annotated RNA on the RNA side, or with Star alignment score less than 255. We generate separate files for reads mapping to exons, intron on UTLs. On the DNA side, we filter reads with botwie2 alignment score less that 40. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "function clean_up_bed_dna {\n",
    "     export LC_COLLATE=\"C\"\n",
    "        fin=\"${1}.bed.gz\" #rna\n",
    "        fout=\"${3}\"\n",
    "        if [ \"${2}\" = \"exons\" ]; then\n",
    "            \n",
    "\n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}(substr($12,4,1)==\"T\" && $10>254 && $5>39){print $1, $2, $3, $4, $5, $6}'> \"${fout}\"\n",
    "\n",
    "        elif [ \"${2}\" = \"introns\" ]; then\n",
    "           \n",
    "\n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}(substr($12,4,1)==\"G\" && $10>254 && $5>39){print $1, $2, $3, $4, $5, $6}'  > \"${fout}\"\n",
    "        else \n",
    "            \n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($10>254 && $5>39){print $1, $2, $3, $4, $5, $6}'  > \"${fout}\"\n",
    "        fi\n",
    "}\n",
    "\n",
    "function clean_up_bed_rna {\n",
    "     export LC_COLLATE=\"C\"\n",
    "        fin=\"${1}.bed.gz\" #rna\n",
    "        fout=\"${3}\"\n",
    "        if [ \"${2}\" = \"exons\" ]; then\n",
    "            \n",
    "\n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && substr($12,4,1)==\"T\" && $10>39){print $1, $2, $3, $4, $5, $6}'> \"${fout}\"\n",
    "\n",
    "        elif [ \"${2}\" = \"introns\" ]; then\n",
    "           \n",
    "\n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && substr($12,4,1)==\"G\" && $10>39){print $1, $2, $3, $4, $5, $6}'  > \"${fout}\"\n",
    "        else \n",
    "            \n",
    "            zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && $10>39){print $1, $2, $3, $4, $5, $6}'  > \"${fout}\"\n",
    "        fi\n",
    "}\n",
    "\n",
    "export -f clean_up_bed_dna\n",
    "export -f clean_up_bed_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "module load system parallel\n",
    "\n",
    "#First cleanup dna.bed files to remove multimappers (Q<40), and split file based on whether the corresponding RNA side of the reads maps to exons, introns. all captures exons and introns together in the same file.\n",
    "parallel --dry-run -j8 'clean_up_bed_dna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna {3} /oak/stanford/groups/astraigh/differentiation_paper_data/data/chardata/dna.blacklisted.Q255Q40.{2}.{3}.bed' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all exons introns\n",
    "                                    \n",
    "#Repeat with reads where the RNA doesn't map to annotated genes but to UTLs. \n",
    "parallel --dry-run -j8 'clean_up_bed_dna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna {3} /oak/stanford/groups/astraigh/differentiation_paper_data/data/chardata/dna.blacklisted.Q255Q40.{2}.intergenic.bed' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all\n",
    "\n",
    "#cleanup rna.bed files to remove multimappers (STAR Q <255) or reads that can not be unambiguously annotated by a signle gene.  \n",
    "parallel --dry-run -j8 'clean_up_bed_rna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/rna {3} /oak/stanford/groups/astraigh/differentiation_paper_data/data/chardata/rna.blacklisted.Q255Q40.{2}.{3}.bed' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all exons introns\n",
    "                                    \n",
    "#repeat with reads where the RNA doesn't map to annotated genes but to UTLs.                                 \n",
    "parallel --dry-run -j8 'clean_up_bed_rna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/rna {3} /oak/stanford/groups/astraigh/differentiation_paper_data/data/chardata/rna.blacklisted.Q255Q40.{2}.intergenic.bed' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate .COOcounts files for cis binding RNAs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`rna.Q255Q40.cis.{exons|introns|all}.COOcounts.tsv` contain the mapping coordiate of the RNA side of each read where the DNA contact is on CIS chromosome. The columns are RNA_chr, RNA_3prime_end, ENSGnumber. They are sorted by ENSGnumber then position. \n",
    "\n",
    "Requires `rna.bed.gz`.\n",
    "\n",
    "We make it for each sample using `rna.bed.gz` file located, with respect to the root of the pipeline output for the specific sample, in the following folders\n",
    "- for the RNA mapping to annotated genes, using the rna.bed.gz file in `pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/`\n",
    "- for the RNA mapping to UTLs, using the rna.bed.gz file in `pairs/gencondeV29_hg38/strngtieCharOnly_hg38/exons/DNAq15-blacklisted_RNAunq/`\n",
    "\n",
    "\n",
    "These files are used to model the CIS contacts and run the generative model for producing predicted ChAR-seq maps.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "function make_COO_rnaCIS {\n",
    "    export LC_COLLATE=\"C\"\n",
    "    fin=\"${1}.bed.gz\" #rna\n",
    "    if [ \"${2}\" = \"exons\" ]; then\n",
    "        fout=\"${1}.Q255Q40.cis.exons.COOcounts.tsv\"\n",
    "    \n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && substr($12,4,1)==\"T\" && $18==1 && $10>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "        \n",
    "    elif [ \"${2}\" = \"introns\" ]; then\n",
    "        fout=\"${1}.Q255Q40.cis.introns.COOcounts.tsv\"\n",
    "    \n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && substr($12,4,1)==\"G\" && $18==1 && $10>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "    else \n",
    "        fout=\"${1}.Q255Q40.cis.all.COOcounts.tsv\"\n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($5>254 && $18==1 && $10>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "    fi\n",
    "}\n",
    "\n",
    "\n",
    "export -f make_COO_rnaCIS\n",
    "\n",
    "#run for reads where the RNA maps to annotated genes, separately for exons, introns, or both exons and introns together in the same file (all)\n",
    "parallel --dry-run -j8 'make_COO_rnaCIS /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/rna {3}' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: exons introns all\n",
    "\n",
    "# repeat for RNAs mapping to UTLs\n",
    "parallel --dry-run -j6 'make_COO_rnaCIS /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/rna {3}' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Generate .COOcounts files for DNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`dna.Q255Q40.{exons|introns|all}.COOcounts.tsv` contains the DNA side mapping position of all contacts from exonic/intronic or total reads. Sorted by ENSG then chr/pos. Of the form\n",
    "```text\n",
    "chr1    27566015        ENSG00000000003.14\n",
    "chr1    27566015        ENSG00000000003.14\n",
    "chr1    46476066        ENSG00000000003.14\n",
    "```\n",
    "\n",
    "Requires `dna.bed.gz`.\n",
    "\n",
    "We make it for each sample using `dna.bed.gz` file located, with respect to the root of the pipeline output for the specific sample, in the following folders\n",
    "- for the RNA mapping to annotated genes, using the dna.bed.gz file in `pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/`\n",
    "- for the RNA mapping to UTLs, using the dna.bed.gz file in `pairs/gencondeV29_hg38/strngtieCharOnly_hg38/exons/DNAq15-blacklisted_RNAunq/`\n",
    "\n",
    "These files are used to ingest the ChARseq data into `chartable` objects using `chartools`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "function make_COO_dna {\n",
    "    export LC_COLLATE=\"C\"\n",
    "    fin=\"${1}.bed.gz\" #dna\n",
    "    if [ \"${2}\" = \"exons\" ]; then\n",
    "        fout=\"${1}.Q255Q40.exons.COOcounts.tsv\"\n",
    "    \n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}(substr($12,4,1)==\"T\" && $10>254 && $5>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "        \n",
    "    elif [ \"${2}\" = \"introns\" ]; then\n",
    "        fout=\"${1}.Q255Q40.introns.COOcounts.tsv\"\n",
    "    \n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}(substr($12,4,1)==\"G\" && $10>254 && $5>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "    else \n",
    "        fout=\"${1}.Q255Q40.all.COOcounts.tsv\"\n",
    "        zcat \"${fin}\" | awk -F $'\\t' 'BEGIN{OFS=FS}($10>254 && $5>39){print $1, $2, $17}' | sort -k3,3 -k1,1 -k2,2n > \"${fout}\"\n",
    "    fi\n",
    "}\n",
    "\n",
    "export -f make_COO_dna\n",
    "\n",
    "parallel -u -t -j8 'make_COO_dna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna {3}' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all exons introns\n",
    "                                    \n",
    "parallel --dry-run -j6 'make_COO_dna /oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries/{1}/data/{2}/pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna {3}' ::: novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun novchar2/CL_2020-05-06_rerun :::+ TFONM2_ES TGNQ5_DE TMF2_Rep2B_ES TMF3_Rep2A_DE ::: all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ingest ChARseq data into chartable objects for data handling in python through the `chartools` package"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load necessary packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from tqdm import tqdm_notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from chartools import charcounter as cc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define where to find the data files for each samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#common root folder for the snakemake pipeline runs for various samples\n",
    "data_root='/oak/stanford/groups/astraigh/differentiation_paper_data/charles/libraries'\n",
    "# subfolder containing the output files of the snakemake preprocessing pipeline for ES_rep1 (sample name TFONM2_ES), DE_rep1 (sample name TGNQ5_DE), \n",
    "NOVA1=\"novchar1/NOVAseq_06-27-2019/CL_2020-05-06_rerun/data\"\n",
    "# subfolder containing the output files of the snakemake preprocessing pipeline for ES_rep2 (sample name TMF2_Rep2B_ES), DE_rep2 (sample name TMF3_Rep2A_DE),\n",
    "NOVA2=\"novchar2/CL_2020-05-06_rerun/data\"\n",
    "\n",
    "NOVA1_smpls=['TFONM2_ES','TGNQ5_DE']\n",
    "NOVA2_smpls=['TMF2_Rep2B_ES','TMF3_Rep2A_DE']\n",
    "libs={}\n",
    "for k in NOVA1_smpls:\n",
    "    libs[k]=os.path.join(data_root,NOVA1, k)\n",
    "for k in NOVA2_smpls:\n",
    "    libs[k]=os.path.join(data_root,NOVA2, k)\n",
    "    \n",
    "#list of all samples to process (in order: ES_rep1, DE_rep1, ES_rep2, DE_rep2)\n",
    "smpls=['TFONM2_ES','TGNQ5_DE','TMF2_Rep2B_ES','TMF3_Rep2A_DE']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Load gene coordinates and chromosome size info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "genes=pd.read_csv('/oak/stanford/groups/astraigh/charseq2.0/genomes/hsapiens/grch38_foralign/ENSG_1_X.csv', index_col=0)\n",
    "\n",
    "chr_df=pd.read_csv('/oak/stanford/groups/astraigh/charseq2.0/genomes/hsapiens/grch38_foralign/resources/chrNameLength_1-22X.txt', names=['chr','len'], sep='\\t', index_col=0)\n",
    "\n",
    "gene_strand_dict={k: (True if v==\"-\" else False) for k, v in genes.strand.to_dict().items()}\n",
    "\n",
    "\n",
    "chr_list=list(chr_df.index)\n",
    "chr_list.sort()\n",
    "chr_to_chrid={v:i for i,v in enumerate(chr_list)}\n",
    "\n",
    "gene_to_chrid=genes['chr'].map(chr_to_chrid).to_dict()\n",
    "\n",
    "for k, v in gene_to_chrid.items():\n",
    "    gene_to_chrid[k]=[v,v]\n",
    "    \n",
    "gene_to_id={k:[i,gene_to_chrid[k][0]] for i, k in enumerate(list(genes.index))}\n",
    "\n",
    "chr_LUT=[\"\"]*len(chr_to_chrid)\n",
    "for k, v in chr_to_chrid.items():\n",
    "    chr_LUT[v]=k\n",
    "    \n",
    "chr_len_vec=chr_df.reindex(chr_LUT).values.flatten()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load coordinates of all UTL genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "intergenes=pd.read_csv('/home/groups/astraigh/differentiation_charseq_paper/notebooks/charles/13_intergenic/data/charOnly/strngtie.1to22X.NOdup.csv', index_col=0)\n",
    "\n",
    "intergenes_strand_dict={k: (True if v==\"-\" else False) for k, v in intergenes.strand.to_dict().items()}\n",
    "\n",
    "intergene_to_chrid=intergenes['chr'].map(chr_to_chrid).to_dict()\n",
    "\n",
    "for k, v in intergene_to_chrid.items():\n",
    "    intergene_to_chrid[k]=[v,v]\n",
    "    \n",
    "intergene_to_id={k:[i,intergene_to_chrid[k][0]] for i, k in enumerate(list(intergenes.index))}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ingest into chartable objects RNA coordinates of the RNA-DNA contact point for each RNA binding to CIS chrosomome.\n",
    "\n",
    "These objects will be used for building generative model for ChAR-seq data and producing expected contact maps. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We load the COO files produces above `rna.Q255Q40.cis.{exons|introns|all}.COOcounts.tsv` into a chartable object and pickle it into `rna.Q255.cis.{exons|introns|all}.chartable.{resolution}.pickle`. Once pickled, these files can be reloaded later. \n",
    "\n",
    "We do this first ror RNAs mapping to annotated genes. We generate separate chartable objects with just the RNAs mapping to exons or introns. We also generate a \"combined\" chartable with all the RNAs mapping to annotated genes, regardless if the mapping is to exons or introns.\n",
    "\n",
    "The pickled chartables are located in:\n",
    "\n",
    "`pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/rna.Q255Q40.cis.{exons|introns|all}.chartable.{resolution}.pickle`\n",
    "\n",
    "Resolution is always 1bp for RNA contact points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i, s in enumerate(smpls):\n",
    "    for t in ['all','exons','introns']:\n",
    "        for j in [1]:\n",
    "            print((s,t,j))\n",
    "            rnaCOO=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/rna.Q255Q40.cis.%s.COOcounts.tsv'%t)\n",
    "            rnaCOOout=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/rna.Q255Q40.cis.%s.chartable.%g.pickle'%(t,j))\n",
    "            tto=cc.Chartable_binned(chr_to_chrid, chr_len_vec, annot_chr_dict=gene_to_id, binsize=j)\n",
    "            tt=tto.copy()\n",
    "            tt.load_raw_counts(rnaCOO, reorganize_genes=False, compress=True, compress_at=None)\n",
    "\n",
    "            with open(rnaCOOout, 'wb') as output:\n",
    "                pickle.dump(tt, output, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We repeat this for RNAs mapping to UTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i, s in enumerate(smpls):\n",
    "    for t in ['all']:\n",
    "        for j in [1]: \n",
    "            print((s,t,j))\n",
    "            rnaCOO=os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/rna.Q255Q40.cis.%s.COOcounts.tsv'%t)\n",
    "            rnaCOOout=os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/rna.Q255Q40.cis.%s.chartable.%g.pickle'%(t,j))\n",
    "            tto=cc.Chartable_binned(chr_to_chrid, chr_len_vec, annot_chr_dict=intergene_to_id, binsize=j)\n",
    "            tt=tto.copy()\n",
    "            tt.load_raw_counts(rnaCOO, reorganize_genes=False, compress=True, compress_at=None)\n",
    "\n",
    "            with open(rnaCOOout, 'wb') as output:\n",
    "                pickle.dump(tt, output, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ingest into chartable objects the DNA contacts for each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`chartables` stores genomic coordinates in compressed format, where only the genomic coordinates of where contacts are observed (typically DpnII sites) are stored. The loci are indexed and serve as columns in the sparse matrix at the core of the chartable objects. Because contacts may be observed at different subets of genomic loci for each sample, the meaning of the column ID could be different for each sample. Because we don't want that, we don't compress all the chartables independently from one another. Instead, we first make a non-compressed version of these tables, (where each column represents a 10bp interval on the genome) `dna.Q255.{exons|introns|all}.chartable.nocompress.10.pickle`. Then we will next compress the data from all samples together, sot that exons/introns/UTLs chartables for all samples have the same compression scheme (meaning each column represents the same genomic coordinate) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Uncompressed chartables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the RNAs mapping to annotated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for i, s in enumerate(smpls):\n",
    "    for t in ['all','exons','introns']:\n",
    "        for j in [10]: #resolution of the binning 10bp\n",
    "            print((s,t,j))\n",
    "            dnaCOO=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.COOcounts.tsv'%t)\n",
    "            dnaCOOout=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(t,j))\n",
    "            tto=cc.Chartable_binned(chr_to_chrid, chr_len_vec, annot_chr_dict=gene_to_id, binsize=j)\n",
    "            tt=tto.copy()\n",
    "            tt.load_raw_counts(dnaCOO, reorganize_genes=False, compress=False, compress_at=None)\n",
    "\n",
    "            with open(dnaCOOout, 'wb') as output:\n",
    "                pickle.dump(tt, output, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For RNAs mapping to UTLs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#intergeninc stuff\n",
    "for i, s in enumerate(smpls):\n",
    "    for t in ['all']:\n",
    "        for j in [10]: #resolution of the binning 10bp\n",
    "            print((s,t,j))\n",
    "            dnaCOO=os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.COOcounts.tsv'%t)\n",
    "            print(dnaCOO)\n",
    "            dnaCOOout=os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(t,j))\n",
    "            tto=cc.Chartable_binned(chr_to_chrid, chr_len_vec, annot_chr_dict=intergene_to_id, binsize=j)\n",
    "            tt=tto.copy()\n",
    "            tt.load_raw_counts(dnaCOO, reorganize_genes=False, compress=False, compress_at=None)\n",
    "\n",
    "            with open(dnaCOOout, 'wb') as output:\n",
    "                pickle.dump(tt, output, pickle.HIGHEST_PROTOCOL)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compress all DNA chartables made above using joint support"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we want to compress the chartable made above for easy loading/handling. We do that by loading the exons/introns and UTLs tables , and then compressing then using the joint support of all three as the compression support. The compressed ones go into --> `dna.Q255.{exons|introns|all}.chartable.compress.10.pickle`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for j in [10]:\n",
    "    alldata={}\n",
    "    for i, s in enumerate(smpls):\n",
    "        print((j, s))\n",
    "        dnae=pickle.load(open(os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(\"exons\",j)),'rb'))\n",
    "        dnai=pickle.load(open(os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(\"introns\",j)),'rb'))\n",
    "        dnaa=pickle.load(open(os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(\"all\",j)),'rb'))\n",
    "        \n",
    "        dnax=pickle.load(open(os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.nocompress.%g.pickle'%(\"all\",j)),'rb'))\n",
    "        hei=dnae.sum(axis=0)\n",
    "        hi=dnai.sum(axis=0)\n",
    "        hx=dnax.sum(axis=0)\n",
    "        ha=dnaa.sum(axis=0)\n",
    "        #hei.counts=hei.counts+hi.counts+ha.counts #+hx.counts WOOOOO\n",
    "        \n",
    "        hei.counts=hei.counts+hi.counts+ha.counts+hx.counts\n",
    "        hei.compress()\n",
    "        dnae.compress(x=hei.x.copy())\n",
    "        dnai.compress(x=hei.x.copy())\n",
    "        dnax.compress(x=hei.x.copy())\n",
    "        dnaa.compress(x=hei.x.copy())\n",
    "        \n",
    "        \n",
    "        dnaCOOouta=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"all\",j))\n",
    "            \n",
    "        with open(dnaCOOouta, 'wb') as output:\n",
    "                pickle.dump(dnaa, output, pickle.HIGHEST_PROTOCOL)\n",
    "        \n",
    "        dnaCOOoute=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"exons\",j))\n",
    "            \n",
    "        with open(dnaCOOoute, 'wb') as output:\n",
    "                pickle.dump(dnae, output, pickle.HIGHEST_PROTOCOL)\n",
    "                \n",
    "        dnaCOOouti=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"introns\",j))\n",
    "            \n",
    "        with open(dnaCOOouti, 'wb') as output:\n",
    "                pickle.dump(dnai, output, pickle.HIGHEST_PROTOCOL)\n",
    "                \n",
    "        dnaCOOoutx=os.path.join(libs[s],'pairs/strngtieCharOnly_hg38/exons/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"all\",j))\n",
    "            \n",
    "        with open(dnaCOOoutx, 'wb') as output:\n",
    "                pickle.dump(dnax, output, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Charseq-ability background\n",
    "\n",
    "The last thing to do is generate the background model used for correcting the generative models and ChARseq contact maps predictions. To generate the background model, we need to combine the trans-binding profiles of all RNAs that have not been annotated as trans-delocalized (see paper)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first generate the background model for exonic RNAs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tscores_exons=pd.read_csv('/home/groups/astraigh/differentiation_charseq_paper/notebooks/charles/17_tscores/data/tscoresAllSamples.exons.tab', sep='\\t', header=0)\n",
    "\n",
    "mrnas_towards_bkg_exons=tscores_exons.loc[(tscores_exons['sample']=='TFONM2') & (tscores_exons['type_coarse']=='protein_coding') &  (tscores_exons['decile']<10)]['ENSG'].to_list()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "start_time": "2023-07-30T00:06:11.014Z"
    }
   },
   "source": [
    "Now for intronic RNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tscores_introns=pd.read_csv('/home/groups/astraigh/differentiation_charseq_paper/notebooks/charles/17_tscores/data/tscoresAllSamples.introns.tab', sep='\\t', header=0)\n",
    "\n",
    "mrnas_towards_bkg_introns=tscores_introns.loc[(tscores_introns['sample']=='TFONM2') & (tscores_introns['type_coarse']=='protein_coding') &  (tscores_introns['decile']<10)]['ENSG'].to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now for all RNAs combined, regardless of whether they map to exons or introns "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mrnas_towards_bkg_all = list(set(mrnas_towards_bkg_exons+mrnas_towards_bkg_introns))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to make the background model by just loading the compressed DNA chartable and summing up all the signal over the model mRNAs. The output pickled model is -->  `bkg.Q255Q40.{exons|introns|all}.chartable.compress.10.pickle`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for j in [10]: #resolution 10bp\n",
    "    alldata={}\n",
    "    for s in smpls: \n",
    "        print((j, s))\n",
    "        dnae=pickle.load(open(os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"exons\",j)),'rb'))\n",
    "        dnai=pickle.load(open(os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/dna.Q255Q40.%s.chartable.compress.%g.pickle'%(\"introns\",j)),'rb'))\n",
    "        \n",
    "    \n",
    "        \n",
    "        sname=s.split(\"_\")[0]\n",
    "        mrnas_towards_bkg_exons=tscores_exons.loc[(tscores_exons['sample']==sname) & (tscores_exons['type_coarse']=='protein_coding') &  (tscores_exons['decile']<10)]['ENSG'].to_list()\n",
    "        mrnas_towards_bkg_introns=tscores_introns.loc[(tscores_introns['sample']==sname) & (tscores_introns['type_coarse']=='protein_coding') &  (tscores_introns['decile']<10)]['ENSG'].to_list()\n",
    "        \n",
    "#         mrnas_towards_bkg_all = list(set(mrnas_towards_bkg_exons+mrnas_towards_bkg_introns))\n",
    "        \n",
    "        dnaec, dnaet, _=dnae.cis_trans()\n",
    "        dnaic, dnait, _=dnai.cis_trans()\n",
    "        #dnaac, dnaat, _=dnaa.cis_trans()\n",
    "        \n",
    "        annote=[k for k in mrnas_towards_bkg_exons if k in dnae.annot_LUT]\n",
    "        annoti=[k for k in mrnas_towards_bkg_introns if k in dnai.annot_LUT]\n",
    "       # annota=[k for k in mrnas_towards_bkg_all if k in dnaa.annot_LUT]\n",
    "        \n",
    "        Be=dnaet.select_annots(annote).sum(axis=0)\n",
    "        Bi=dnait.select_annots(annoti).sum(axis=0)\n",
    "       # Ba=dnaat.select_annots(annota).sum(axis=0)\n",
    "        \n",
    "        Bei=Be.copy()\n",
    "        Bei.counts=Bei.counts+Bi.counts\n",
    "        \n",
    "        \n",
    "        bout=os.path.join(libs[s],'pairs/gencondeV29_hg38/all/filtered/DNAq15-blacklisted_RNAunq/bkg.Q255Q40.chartable.compress.%g.pickle'%j)\n",
    "            \n",
    "        with open(bout, 'wb') as output:\n",
    "            pickle.dump({'exons':Be,'introns':Bi, 'all':Bei}, output, pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_clean",
   "language": "python",
   "name": "scanpy_clean"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
